pe.c

```diff
@@ -39,6 +39,130 @@
 })
 #define check_size(d, ds, h, hs) check_size_line(d, ds, h, hs, __LINE__)
 
+static EFI_STATUS
+_do_sha256_sum(void *addr, UINTN size, UINT8 *digest)
+{
+	unsigned int sha256ctxsize;
+	void *sha256ctx = NULL;
+
+	sha256ctxsize = Sha256GetContextSize();
+	sha256ctx = AllocateZeroPool(sha256ctxsize);
+	if (sha256ctx == NULL)
+		return EFI_OUT_OF_RESOURCES;
+
+	Sha256Init(sha256ctx);
+	Sha256Update(sha256ctx, addr, size);
+	Sha256Final(sha256ctx, digest);
+
+	FreePool(sha256ctx);
+	return EFI_SUCCESS;
+}
+
+struct shim_section_cache_entry {
+	EFI_HANDLE parent_image_handle;
+	UINT8 section_name[8];
+	void *addr;
+	UINTN size;
+	/*
+	 * Since this is all internal and there's no API access to it, this is
+	 * currently always sha256 and can be updated as needed.
+	 */
+	UINT8 digest[32];
+};
+
+static struct shim_section_cache_entry *section_cache = NULL;
+static UINTN num_section_cache_entries = 0;
+
+static EFI_STATUS
+cache_section(EFI_HANDLE parent_image_handle, UINT8 section_name[8],
+	      void *addr, UINTN size)
+{
+	struct shim_section_cache_entry *new_section_cache = NULL;
+	struct shim_section_cache_entry *entry = NULL;
+	size_t oscsz = num_section_cache_entries * sizeof (*new_section_cache);
+	size_t nscsz = oscsz + sizeof (*new_section_cache);
+	EFI_STATUS efi_status;
+
+	new_section_cache = AllocateZeroPool(nscsz);
+	if (!new_section_cache)
+		return EFI_OUT_OF_RESOURCES;
+
+	if (section_cache) {
+		CopyMem(new_section_cache, section_cache, oscsz);
+		FreePool(section_cache);
+	}
+	section_cache = new_section_cache;
+	entry = &section_cache[num_section_cache_entries];
+
+	entry->parent_image_handle = parent_image_handle;
+	CopyMem(entry->section_name, section_name, sizeof(entry->section_name));
+	entry->addr = addr;
+	entry->size = size;
+
+	efi_status = _do_sha256_sum(addr, size, entry->digest);
+	if (EFI_ERROR(efi_status)) {
+		ZeroMem(entry, sizeof (*entry));
+		return efi_status;
+	}
+	return EFI_SUCCESS;
```

Comment:
`num_section_cache_entries` needs to be incremented before returning, otherwise the validate/flush helpers below will not be able to iterate over the cache

Reply:
Fixed in my current tree.